pkg/proxy/iptables/proxier.go

```diff
@@ -1042,80 +1045,108 @@ func (proxier *Proxier) syncProxyRules() {
 				proxier.natChains.WriteBytes(chain)
 			} else {
-				proxier.natChains.Write(utiliptables.MakeChainLine(policyClusterChain))
+				proxier.natChains.Write(utiliptables.MakeChainLine(clusterPolicyChain))
 			}
-			activeNATChains[policyClusterChain] = true
+			activeNATChains[clusterPolicyChain] = true
 		}
 
-		if hasEndpoints && svcInfo.ExternallyAccessible() && svcInfo.NodeLocalExternal() {
-			if chain, ok := existingNATChains[svcXlbChain]; ok {
+		// Declare the localPolicyChain if needed.
+		if hasEndpoints && svcInfo.UsesLocalEndpoints() {
+			if chain, ok := existingNATChains[localPolicyChain]; ok {
 				proxier.natChains.WriteBytes(chain)
 			} else {
-				proxier.natChains.Write(utiliptables.MakeChainLine(svcXlbChain))
+				proxier.natChains.Write(utiliptables.MakeChainLine(localPolicyChain))
 			}
-			activeNATChains[svcXlbChain] = true
+			activeNATChains[localPolicyChain] = true
+		}
 
-			// The XLB chain redirects all pod -> external VIP
-			// traffic to the Service's ClusterIP instead. This happens
-			// whether or not we have local endpoints; only if localDetector
-			// is implemented
-			if proxier.localDetector.IsImplemented() {
-				proxier.natRules.Write(
-					"-A", string(svcXlbChain),
-					"-m", "comment", "--comment",
-					`"Redirect pods trying to reach external loadbalancer VIP to clusterIP"`,
-					proxier.localDetector.IfLocal(),
-					"-j", string(policyClusterChain))
+		// If any "external" destinations are enabled, set up external traffic
+		// handling.  All captured traffic for all external destinations should
+		// jump to externalTrafficChain, which will handle some special-cases
+		// and then jump to externalPolicyChain.
+		if hasEndpoints && svcInfo.ExternallyAccessible() {
+			if chain, ok := existingNATChains[externalTrafficChain]; ok {
+				proxier.natChains.WriteBytes(chain)
+			} else {
+				proxier.natChains.Write(utiliptables.MakeChainLine(externalTrafficChain))
 			}
+			activeNATChains[externalTrafficChain] = true
 
-			// Next, redirect all src-type=LOCAL -> LB IP to the service chain
-			// for externalTrafficPolicy=Local This allows traffic originating
-			// from the host to be redirected to the service correctly,
-			// otherwise traffic to LB IPs are dropped if there are no local
-			// endpoints.
-			proxier.natRules.Write(
-				"-A", string(svcXlbChain),
-				"-m", "comment", "--comment", fmt.Sprintf(`"masquerade LOCAL traffic for %s LB IP"`, svcNameString),
-				"-m", "addrtype", "--src-type", "LOCAL",
-				"-j", string(KubeMarkMasqChain))
-			proxier.natRules.Write(
-				"-A", string(svcXlbChain),
-				"-m", "comment", "--comment", fmt.Sprintf(`"route LOCAL traffic for %s LB IP to service chain"`, svcNameString),
-				"-m", "addrtype", "--src-type", "LOCAL",
-				"-j", string(policyClusterChain))
+			if !svcInfo.ExternalPolicyLocal() || !proxier.localDetector.IsImplemented() {
+				// If we have are using non-local endpoints (or if we can't
```

Comment:
oh, since you have to re-push anyway:

```suggestion
				// If we are using non-local endpoints (or if we can't
```

Reply:
ACK